vm/systemSmartContracts/staking.go

```diff
@@ -999,6 +1008,49 @@ func (r *stakingSC) createWaitingListKey(blsKey []byte) []byte {
 	return []byte(waitingElementPrefix + string(blsKey))
 }
 
+func (r *stakingSC) switchJailedWithWaiting(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, r.endOfEpochAccessAddr) {
+		r.eei.AddReturnMessage("switchJailedWithWaiting function not allowed to be called by address " + string(args.CallerAddr))
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) != 1 {
+		return vmcommon.UserError
+	}
+
+	registrationData, err := r.getOrCreateRegisteredData(args.Arguments[0])
+	if err != nil {
+		return vmcommon.UserError
+	}
+	if len(registrationData.RewardAddress) == 0 {
+		return vmcommon.UserError
+	}
+	if !registrationData.Staked {
+		return vmcommon.UserError
+	}
+	if registrationData.Jailed {
+		r.eei.AddReturnMessage(vm.ErrBLSPublicKeyAlreadyJailed.Error())
+		return vmcommon.UserError
+	}
+	err = r.moveFirstFromWaitingToStakedIfNeeded(args.Arguments[0])
+	if err != nil {
+		r.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	r.removeFromStakedNodes()
+	registrationData.Staked = false
```

Comment:
registration data will contain the information related to the waiting or to the jailed node?

Reply:
it already has.